core/src/main/scala/chisel3/Data.scala

```diff
@@ -595,8 +615,8 @@ abstract class Data extends HasId with NamedComponent with SourceInfoDoc {
         }
       }
 
-    val leftType = if (this.hasBinding) this.cloneType else this
-    val rightType = if (that.hasBinding) that.cloneType else that
+    val leftType = this.cloneTypeFull //if (this.hasBinding) this.cloneType else this
+    val rightType = that.cloneTypeFull //if (that.hasBinding) that.cloneType else that
```

Comment:
```suggestion
    val leftType = this.cloneTypeFull
    val rightType = that.cloneTypeFull
```

I don't know that I fully understand this change (I get as far as "copy over probe-ness of top-level of each"), but just checking could/should/might the clone still be gated on `hasBinding`?

Reply:
yeah this change I am really shocked does not break anything. Because i would imagine if you have:

class Producer extends Module{
  val out = IO(Output(Bool()))
}

class Consumer extends Module {
  val in = IO(Input(Bool())
}

these will now fail type equivalence...? But perhaps we don't actually run this check for the `<>` or `:<>=` top level operators?

Reply:
I don't think this function checks directions so I don't think it will fail in that case.

My question is, do we even need to clone? Why not just check the Data as they are? I wrote this code a year ago and can't remember any justification, but I also adapted it from other code that may have done it for no reason. Worth trying to remove the cloning completely.